### PR TITLE
Always allow tensor metadata changes

### PR DESCRIPTION
**Pull request by bdhirsh**
Make it so that it is valid to set metadata after detach calls, like `x.detach().resize_(...)`.

This technically lifts some restrictions around `.data`. This PR means that you can now technically call `x.data.resize_(...)`, which can now directly resize `x` instead of erroring.

My understanding: Before the tensor-variable merge, when `x` and `x.data` were really different tensors, you could resize `x.data` independently of `x`, and during the merge, this error was added to avoid silent confusing behavior changes.

It was agreed that this error has been around long enough (several years) that it's acceptable to drop.  cc @albanD @ezyang.

(Ed already had a prototype PR [here](https://github.com/pytorch/pytorch/pull/83545) - I ended up making one to try to slog through test failures).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83701
* #83626
* #83542
* __->__ #83590

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -7212,22 +7212,8 @@ def test_has_internal_overlap(self):
...
-                    "set_storage is not allowed on a Tensor created from .data or .detach()"):
-                t.set_()
-            with self.assertRaisesRegex(
-                    RuntimeError,
-                    "set_storage_offset is not allowed on a Tensor created from .data or .detach()"):
-                t.set_(t.storage(), 0, t.size(), list(t.stride()))
-
-        do_test(torch.tensor([[1, 2]]).data)
-        do_test(torch.tensor([[1, 2]]).detach())
+        a = torch.ones(2, 3)
```
Comment (albanD) [inline test/test_torch.py:7215]: The whole test can go I guess?

**Review by ezyang — APPROVED** [review]
